svm/src/nonce_info.rs

```diff
@@ -14,10 +23,39 @@ pub struct NoncePartial {
     account: AccountSharedData,
 }
 
+#[derive(Error, Debug, PartialEq)]
+pub enum AdvanceNonceError {
+    #[error("Invalid account")]
+    Invalid,
+    #[error("Uninitialized nonce")]
+    Uninitialized,
+}
+
 impl NoncePartial {
     pub fn new(address: Pubkey, account: AccountSharedData) -> Self {
         Self { address, account }
     }
+
+    // Advance the stored blockhash to prevent fee theft by someone
+    // replaying nonce transactions that have failed with an
+    // `InstructionError`.
+    pub fn try_advance_account(
```

Comment:
nit: naming `try_advance_nonce` it looks like _advance_ takes `nonce` as its objective more often in the code base?

Reply:
nit: I'm still inclined to panic here when called twice due to my personal preference, even after sleeping. but i know this is quite imho thing. ;)

Reply:
it's quite unfortunate for us not to be able to consume `self` here easily due to surrounding code. ;)

Reply:
What's your concern about it being called twice? The method is idempotent, isn't it?

Reply:
> The method is idempotent, isn't it?

It is now. however, it might not be in the very long future theoretically. (I know this is quite unrealistic though.)

> What's your concern about it being called twice?

The above thing said, I'm fond of putting panics into code as a habit (unless significant runtime overhead is incurred), which triggers on any unintended runtime usage, if type system can't prevent that. (i.e. rewriting `NoncePartial::try_advance_account` to consume `self` isn't easy in this case).

if the same nonce account is rolled back twice, it sounds like there's odd behavior or a bug to me. put differently, if we're enough confident of the code, put panic and take blame of loa, should that happen, instead of lof.

as a different example, i always try to assert that `hash_map.insert(..)` did removed the existing entry or not, if only either of the outcome can be proven.

yet another subjective opinion is that it's easy for future engineers to hack/read on code which fails loudly in any misuse.

Reply:
Ok I understand your concern now, I also like to be defensive when I can so thanks for the suggestion. I will go with the change as it stands right now but I will think about how to consume `RollbackAccounts` when committing transactions so that we can consume the `NoncePartial` here.